teps/0011-redirecting-step-output-streams.md

```diff
@@ -0,0 +1,219 @@
+
+## Requirements
+
+* Add new fields for a step to specify paths to redirect stdout/stderr to.
+* Users should be able to observe the output streams through Pod logs even if `StdoutPath` or `StderrPath` is specified. In other words, output streams should be duplicated instead of simply redirected.
+* Clearly documents the restrictions when `StdoutPath` and `StderrPath` are set to Task result paths, and encourage Task authors to set the paths to workspace paths, especially if they want to exchange large amount of data between Tasks.
+* Provide examples to use step redirection for the use cases mentioned in [Motivation](#motivation).
+
+## Proposal
+
+### Option 1: Allowing Users to Specify Redirection Paths
+
+The following new fields will be added to the `Step` struct:
+
+```go
+type Step struct {
+    ...
+    // Path to duplicate stdout stream to on container's local filesystem.
+    // +optional
+    StdoutPath string `json:"stdout_path,omitempty"`
+    // Path to duplicate stderr stream to on container's local filesystem.
+    // +optional
+    StderrPath string `json:"stderr_path,omitempty"`
+}
+```
+
+Once `StdoutPath` or `StderrPath` is specified, the corresponding output stream will be duplicated to both the given file and the standard output stream of the container, so users can still view the output through the Pod log API. If both `StdoutPath` and `StderrPath` are set to the same value, the two streams will be written to two file descriptors referring to the same file description, so outputs from both stream will be interleaved in the same file, but there will be no ordering guarantee on the data.  If multiple steps' `StdoutPath` fields are set to the same value, the file content will be overwritten by the last outputting step.
+
+Variable substitution will be applied to the new fields, so one could specify `$(results.<name>.path)` to the `StdoutPath` field to extract the stdout of a step into a Task result. No new variable substitution for accessing the values of `StdoutPath` and `StderrPath` fields will be provided so variable substitution can remain single-pass.
+
+#### Example Usage
+
+Redirecting stdout of `boskosctl` to `jq` and publish the resulting `project-id` as a Task result:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: boskos-acquire
+spec:
+  results:
+  - name: project-id
+  steps:
+  - name: boskosctl
+    image: gcr.io/k8s-staging-boskos/boskosctl
+    args:
+    - acquire
+    - --server-url=http://boskos.test-pods.svc.cluster.local
+    - --owner-name=christie-test-boskos
+    - --type=gke-project
+    - --state=free
+    - --target-state=busy
+    stdoutPath: /data/boskosctl-stdout
+    volumeMounts:
+    - name: data
+      mountPath: /data
+  - name: parse-project-id
+    image: imega/jq
+    args:
+    - -r
+    - .name
+    - /data/boskosctl-stdout
+    stdout: $(results.project-id.path)
```

Comment:
also, minor nit: update this to `stdoutPath` to match new syntax.

Reply:
Hmm great point. yeah that could lead to some UX problem. Maybe adding a trim option to `StdoutConfig`? I'm not sure if that's a good idea though, as it sounds like adding extra nonessential processing magic to entrypoint.

Reply:
Hmm instead of trimming spaces when generating stdout, maybe it's better to trim spaces when writing results to the termination path? For example, we could add a trim option in result declaration.

Reply:
Ah, yeah, agree that sounds like a more broadly useful approach!

Reply:
TEP to trim trailing whitespace from results: https://github.com/tektoncd/community/pull/223

Reply:
Quick update, I don't think we should automatically trim whitespace, discussion in https://github.com/tektoncd/community/pull/223 if you are interested

> I'm not sure if jq will include a newline in its output here but if it did then a subsequent task using the project id may likewise have trouble with the result when using it as a param.

in the boskos example specifically note that:

1. the output is not being used _as-is_ as the result, it is being transformed with jq (i.e. a trailing newline in the output from boskos has no effect on the final result)
2. adding `-j` to the jq command strips the final newline (https://github.com/tektoncd/catalog/blob/master/task/boskos-acquire/0.1/boskos-acquire.yaml#L42)
3. jq also needs the `-r` flag otherwise the output is surrounded by quotation marks - which begs the question, do we also need to strip quotation marks from output? i'd rather avoid this question entirely and leave it up to the task author